private/waiter/waiter.go

```diff
@@ -61,20 +61,56 @@ func (w *Waiter) Wait() error {
 					break
 				}
 				result = true
-				for _, val := range vals {
-					if !awsutil.DeepEqual(val, a.Expected) {
-						result = false
-						break
+				if exp, ok := a.Expected.(string); ok {
```

Comment:
The change to be type specific is a find change instead of reflection, but an `a.Expected` can be many more values than string and int. It can actually be any value. This is the reason `awsutil.DeepEqual` is used. The service's waiter will define a field that indicates a specific state of the waiter has occurred. The type of this field is pretty open ended. Such as the `a.Expected` could contain complex value such as structs, maps or lists. 

I don't think off hand there are any that actually use complex types, so a primitive type matching might be used without reflection. Though at any time a service's waiter could use a complex type, so the SDK would need to use the DeepEquals as a fallback.

In either case any functionality like this should be in its own independent function so its not duplicated through the code, and can be tested.